File: go/341.flatten-nested-list-iterator.go
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * type NestedInteger struct {
 * }
 *
 * // Return true if this NestedInteger holds a single integer, rather than a nested list.
 * func (this NestedInteger) IsInteger() bool {}
 *
 * // Return the single integer that this NestedInteger holds, if it holds a single integer
 * // The result is undefined if this NestedInteger holds a nested list
 * // So before calling this method, you should have a check
 * func (this NestedInteger) GetInteger() int {}
 *
 * // Set this NestedInteger to hold a single integer.
 * func (n *NestedInteger) SetInteger(value int) {}
 *
 * // Set this NestedInteger to hold a nested list and adds a nested integer to it.
 * func (this *NestedInteger) Add(elem NestedInteger) {}
 *
 * // Return the nested list that this NestedInteger holds, if it holds a nested list
 * // The list length is zero if this NestedInteger holds a single integer
 * // You can access NestedInteger's List element directly if you want to modify it
 * func (this NestedInteger) GetList() []*NestedInteger {}
 */

type NestedIterator struct {
    list []*NestedInteger   // containing list
    cursor int              // cursor points to next item to access
    itor *NestedIterator    // if the item that cursor points to is a list, itor is
                            // the iterator of that list
}

func Constructor(nestedList []*NestedInteger) *NestedIterator {
    return &NestedIterator{
        list: nestedList,
        cursor: 0,
        itor: nil,
    }
}

func (this *NestedIterator) Next() int {
    if this.list[this.cursor].IsInteger() {
        // next item is integer
        value := this.list[this.cursor].GetInteger()
        // advance the cursor for next iteration
        this.cursor++
        return value
    } else {
        // next item is in subordinate list
        return this.itor.Next()
    }
}

func (this *NestedIterator) HasNext() bool {
    // cursor exceeds list length
    if this.cursor >= len(this.list) {
        return false
    }
    if this.list[this.cursor].IsInteger() {
        // next item to access is an integer
        return true
    } else {
        // next item to access is a subordinate list
        if this.itor == nil {
            // very first time to access the subordinate list
            // initialize iterator for that list
            this.itor = Constructor(this.list[this.cursor].GetList())
        }
        if this.itor.HasNext() {
            // subordinate list has next
            return true
        } else {
            // subordinate list has no next
            // reset iterator of last subordinate list
            this.itor = nil
            // try the next item in current list
            this.cursor++
            return this.HasNext()
        }
    }
}

File: csharp/341.flatten-nested-list-iterator.cs
/*
 * @lc app=leetcode id=341 lang=csharp
 *
 * [341] Flatten Nested List Iterator
 */

// @lc code=start
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * interface NestedInteger {
 *
 *     // @return true if this NestedInteger holds a single integer, rather than a nested list.
 *     bool IsInteger();
 *
 *     // @return the single integer that this NestedInteger holds, if it holds a single integer
 *     // Return null if this NestedInteger holds a nested list
 *     int GetInteger();
 *
 *     // @return the nested list that this NestedInteger holds, if it holds a nested list
 *     // Return null if this NestedInteger holds a single integer
 *     IList<NestedInteger> GetList();
 * }
 */

using System;
using System.Collections.Generic;

public class NestedIterator 
{
    private IEnumerator<NestedInteger> enumerator;

    private NestedIterator currentNestedIterator;

    private bool first;

    public NestedIterator(IList<NestedInteger> nestedList) 
    {
        this.enumerator = nestedList.GetEnumerator();
        this.currentNestedIterator = null;
        this.first = true;
    }

    public bool HasNext() 
    {
        if (this.first)
        {
            // the very first call to HasNext, move to next.
            this.first = false;
            bool hasNext = this.enumerator.MoveNext();
            if (!hasNext)
            {
                return false;
            }
        }
        else
        {
            if (this.enumerator.Current.IsInteger())
            {
                // current cursor points to a integer, move to next.
                // could be merged into the previous 'first' case
                bool hasNext = this.enumerator.MoveNext();
                if (!hasNext)
                {
                    return false;
                }
            }
            else
            {
                // current cursor points to a nested list
                // try to advance the inner cursor
                bool hasNestedNext = this.currentNestedIterator.HasNext();
                if (hasNestedNext)
                {
                    return true;
                }

                // or move to next element at current level if inner move fails.
                bool hasNext = this.enumerator.MoveNext();
                if (!hasNext)
                {
                    return false;
                }
            }
        }

        // current (level) cursor moved to next if code doesn't bait out eariler.
        // we need to inspect next element until there is a concrete integer(integer or nested list with available next element).
        while (true)
        {
            if (this.enumerator.Current.IsInteger())
            {
                return true;
            }

            this.currentNestedIterator = new NestedIterator(this.enumerator.Current.GetList());
            bool hasNestedNext = this.currentNestedIterator.HasNext();
            if (hasNestedNext)
            {
                return true;
            }

            bool hasNext = this.enumerator.MoveNext();
            if (!hasNext)
            {
                return false;
            }
        }
    }

    public int Next() 
    {
        if (this.enumerator.Current.IsInteger())
        {
            return this.enumerator.Current.GetInteger();
        }
        else 
        {
            return this.currentNestedIterator.Next();
        }
    }
}

/**
 * Your NestedIterator will be called like this:
 * NestedIterator i = new NestedIterator(nestedList);
 * while (i.HasNext()) v[f()] = i.Next();
 */
// @lc code=end

